### ISSUE TITLE
proposal: extend use_build_context_synchronously to consider Riverpod's WidgetRef

### ISSUE THREAD

**Issue by rrousselGit** [dart-lang/linter#3419]
Hello! I'm not sure whether this is reasonable, but it's worth asking.

Long story short, [Riverpod](https://pub.dev/packages/flutter_riverpod) would like to define a lint strictly identical to `use_build_context_synchronously` but for its `WidgetRef` object.

The lint would be identical in every point, such that BuildContext/WidgetRef can constantly be interchanged. As a matter of fact, WidgetRef _is_ a BuildContext. It uses a different interface to cause less confusion and allows folks using package:provider to smoothly migrate to package:riverpod.

As such, would it be reasonable to update use_build_context_synchronously to handle Riverpod's WidgetRef too?

Such that we have:

```dart
void fn() async {
  WidgetRef ref;

  await something;
  ref.anything();
  ^^^^^^^^^^ use_build_context_synchronously

  if (mounted) {
    ref.anything(); // ok
  }
```

I understand that this repository probably wants to be clean of third-party package logic. 

If necessary, maybe a solution could be to add a custom package:meta annotation that both WidgetRef and BuildContext would use (`@shouldUseSynchronously` or `@isBuildContext`?)

**Comment by rrousselGit** [comment]
Yes, this applies to WidgetRef too.

**Comment by rrousselGit** [comment]
It's probably worth noting that WidgetRef is a class that interacts with BuildContext (hence why the rules should be respected)

Such that we have:

```dart
abstract class WidgetRef {
  void doSomething() {
    Object.of(context).doSomething();
  }
}
```


So maybe what makes sense would be to annotate the functions of WidgetRef, not WidgetRef itself. Maybe something like `@useBuildContext`?

```dart
abstract class WidgetRef {
  @useBuildContext;
  void doSomething();
}
```

**Comment by bwilkerson** [comment]
Definitely worth considering.

Would this be useful for other users? I don't know Flutter well enough to know whether users might write methods similar to `doSomething` or whether that's unlikely to occur very often.

**Comment by rrousselGit** [comment]
It's probably not very common. Most cases would end-up passing the `context` variable as parameter to their function

The case described here is different because it already has access to the context, so passing it is redundant

**Comment by srawlins** [comment]
CC @goderbauer any strong opinion?

**Comment by rrousselGit** [comment]
For what it's worth, I'm working on riverpod-specific lint-rules using analyzer_plugin

So I could implement this on my own if necessary. Although the annotation approach here would make my like a lot easier
